### PR TITLE
[ENH]: Smoothing images as a preprocessing step

### PR DESCRIPTION
**Pull request by synchon**
### Are you requiring a new dataset or marker?

- [X] I understand this is not a marker or dataset request

### Which feature do you want to include?

One standard preprocessing step in a lot of MRI analyses (that is not applied by fMRIprep) is smoothing images after confound regression. It would be good to have this as an additional inbuilt preprocessing step.

### How do you imagine this integrated in junifer?

The easiest way will probably be using the [nilearn function](https://nilearn.github.io/dev/modules/generated/nilearn.image.smooth_img.html) for smooting images, although one can also use [AFNI](http://andysbrainblog.blogspot.com/2012/06/smoothing-in-afni.html), but I dont expect that to yield very different results.

### Do you have a sample code that implements this outside of junifer?

_No response_

### Anything else to say?

_No response_

### CODE REVIEW

**Comment by synchon** [comment]
> @LeSasse @fraimondo I think we can have either (i) 1 class handling 1/2/3 "backends" (nilearn, AFNI, FSL) or (ii) 3 classes named accordingly. Having 1 class will of course bloat it but keep it in one piece. I personally prefer (ii) but if (i) is better from usage POV, I don't mind it either.

Since, there's no agreement or disagreement, I'll proceed with (ii).

**Comment by fraimondo** [comment]
If there is no "common" API that the different "backends" use, then having one single class can be difficult (e.g. yield too many optional parameters conditional to the backend).

In that case, I would KISS with (ii).

**Comment by fraimondo** [comment]
For me, different backends should not alter the result, rather than the kind of tool used. If the result is not the same, then it should not be a backend, but a different class.

**Comment by synchon** [comment]
> For me, different backends should not alter the result, rather than the kind of tool used. If the result is not the same, then it should not be a backend, but a different class.

After looking at the implementations, they don't exactly do the same thing in the sense that they have different parameters which affect the result so having distinct classes make sense.

**Comment by LeSasse** [comment]
> > @LeSasse @fraimondo I think we can have either (i) 1 class handling 1/2/3 "backends" (nilearn, AFNI, FSL) or (ii) 3 classes named accordingly. Having 1 class will of course bloat it but keep it in one piece. I personally prefer (ii) but if (i) is better from usage POV, I don't mind it either.
> 
> Since, there's no agreement or disagreement, I'll proceed with (ii).

sounds good to me

**Comment by synchon** [comment]
@LeSasse Whenever you feel like, you can give the preprocessors a try.

**Comment by LeSasse** [comment]
> @LeSasse Whenever you feel like, you can give the preprocessors a try.

Will likely get to it next week.

**Comment by synchon** [comment]
@LeSasse This PR works for you right?

**Review by LeSasse — APPROVED** [review]
LGTM

File: junifer/preprocess/smoothing/_nilearn_smoothing.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    This class uses :func:`nilearn.image.smooth_img` to smooth image(s).
+
+    """
+
+    _DEPENDENCIES: ClassVar[Set[str]] = {"nilearn"}
+
+    def preprocess(
+        self,
+        data: "Nifti1Image",
+        fwhm: Union[int, float, ArrayLike, Literal["fast"], None],
```
[the remark this answers is not in the document]
Comment (LeSasse) [reply junifer/preprocess/smoothing/_nilearn_smoothing.py:39]: ah thanks, that makes sense!

**Pull request merged**